source/plugins/ruby/arc_k8s_cluster_identity.rb

```diff
@@ -0,0 +1,210 @@
+    @kube_api_server_url = get_kube_api_server_url
+    @http_client = get_http_client
+    @service_account_token = get_service_account_token
+    @log.info "initialize complete @ #{Time.now.utc.iso8601}"
+  end
+
+  def get_cluster_identity_token()
+    begin
+      # get the cluster msi identity token either if its empty or near expirty. Token is valid 24 hrs.
+      if @cached_access_token.to_s.empty? || (Time.now + 60 * 60 > @token_expiry_time) # Refresh token 1 hr from expiration
+        # renew the token if its near expiry
+        if !@cached_access_token.to_s.empty? && (Time.now + 60 * 60 > @token_expiry_time)
+          @log.info "renewing the token since its near expiry @ #{Time.now.utc.iso8601}"
+          renew_near_expiry_token
+          # sleep 60 seconds to get the renewed token  available
+          sleep 60
+        end
+        @log.info "get token reference from crd @ #{Time.now.utc.iso8601}"
+        tokenReference = get_token_reference_from_crd
+        if !tokenReference.nil? && !tokenReference.empty?
+          @token_expiry_time = Time.parse(tokenReference["expirationTime"])
+          token_secret_name = tokenReference["secretName"]
+          token_secret_data_name = tokenReference["dataName"]
+          # get the token from secret
+          @log.info "get token from secret @ #{Time.now.utc.iso8601}"
+          token = get_token_from_secret(token_secret_name, token_secret_data_name)
+          if !token.nil?
+            @cached_access_token = token
+          else
+            @log.warn "got token nil from secret: #{@token_secret_name}"
+          end
+        else
+          @log.warn "got token reference either nil or empty"
+        end
+      end
+    rescue => err
+      @log.warn "get_cluster_identity_token failed: #{err}"
+    end
+    return @cached_access_token
+  end
+
+  private
+
+  def get_token_from_secret(token_secret_name, token_secret_data_name)
+    token = nil
+    begin
+      secret_request_uri = @@secret_resource_uri_template % {
+        kube_api_server_url: @kube_api_server_url,
+        cluster_identity_token_secret_namespace: @@cluster_identity_token_secret_namespace,
+        token_secret_name: token_secret_name,
+      }
+      get_request = Net::HTTP::Get.new(secret_request_uri)
+      get_request["Authorization"] = "Bearer #{@service_account_token}"
+      @log.info "Making GET request to #{secret_request_uri} @ #{Time.now.utc.iso8601}"
+      get_response = @http_client.request(get_request)
+      @log.info "Got response of #{get_response.code} for #{secret_request_uri} @ #{Time.now.utc.iso8601}"
+      if get_response.code.to_i == 200
+        token_secret = JSON.parse(get_response.body)["data"]
+        cluster_identity_token = token_secret[token_secret_data_name]
+        token = Base64.decode64(cluster_identity_token)
+      end
+    rescue => err
+      @log.warn "get_token_from_secret API call failed: #{err}"
```

Comment:
You can send this to AI to begin with to identify any issues you might encounter

Reply:
> You can send this to AI to begin with to identify any issues you might encounter

> added to capture exceptions in AI.